### PR TITLE
Develop pre-processor for naive SQ8 [MOD-9238]

### PR DESCRIPTION
**Pull request by dor-forer**
**Describe the changes in the pull request**

A clear and concise description of what the PR is solving.

**Which issues this PR fixes**
1. #...
2. MOD...

**Main objects this PR modified**
1. ...
2. ...

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: tests/unit/test_hnsw_tiered.cpp [hunk 1]
```diff
@@ -4204,6 +4204,17 @@ class PreprocessorDoubleValue : public PreprocessorInterface {
...
         : PreprocessorInterface(allocator), dim(dim) {}
+
+    void preprocess(const void *original_blob, void *&storage_blob, void *&query_blob,
+                    size_t &storage_blob_size, size_t &query_blob_size,
+                    unsigned char alignment) const override {
+        // This assert makes sure the current use of the preprocessor is valid,
+        // i.e., both blobs are of the same size.
+        // In order to use different sizes, the preprocessor should be modified.
+        assert(storage_blob_size == query_blob_size);
+        preprocess(original_blob, storage_blob, query_blob, storage_blob_size, alignment);
```
[the remark this answers is not in the document]
Comment (dor-forer) [reply tests/unit/test_hnsw_tiered.cpp:4215]: I see that storage_blob is never updated. So does it affect it?

File: tests/unit/test_components.cpp [hunk 2]
```diff
@@ -945,3 +985,122 @@ TEST(PreprocessorsTest, Int8NormalizeThenIncreaseSize) {
...
+    auto multiPPContainer =
+        MultiPreprocessorsContainer<float, n_preprocessors>(allocator, no_alignment);
+    multiPPContainer.addPreprocessor(quant_preprocessor);
+
+    auto query_blob = multiPPContainer.preprocessQuery(original_blob, original_blob_size);
+    ASSERT_EQ(query_blob.get(), original_blob); // same memory address
+    // Verify content is unchanged
+    EXPECT_NO_FATAL_FAILURE(
+        CompareVectors<float>(static_cast<const float *>(query_blob.get()), original_blob, dim));
+}
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/unit/test_components.cpp]: not supported

File: tests/unit/test_components.cpp [hunk 3]
```diff
@@ -945,3 +985,122 @@ TEST(PreprocessorsTest, Int8NormalizeThenIncreaseSize) {
                                                        final_blob_bytes_count));
     }
 }
+// TODO: test edge case where all entries equal.
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/unit/test_components.cpp]: yes toda

File: tests/unit/test_components.cpp [hunk 4]
```diff
@@ -143,6 +165,12 @@ class DummyMixedPreprocessor : public PreprocessorInterface {
                            DataType value_to_add_storage, DataType value_to_add_query)
         : PreprocessorInterface(allocator), value_to_add_storage(value_to_add_storage),
           value_to_add_query(value_to_add_query) {}
+    void preprocess(const void *original_blob, void *&storage_blob, void *&query_blob,
+                    size_t &storage_blob_size, size_t &query_blob_size,
+                    unsigned char alignment) const override {
+        preprocess(original_blob, storage_blob, query_blob, storage_blob_size, alignment);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/unit/test_components.cpp]: this test is anyway redundant

File: src/VecSim/spaces/computer/preprocessor_container.h [hunk 5]
```diff
@@ -168,10 +168,17 @@ MultiPreprocessorsContainer<DataType, n_preprocessors>::preprocess(const void *o
...
+    size_t storage_blob_size = input_blob_size;
+    size_t query_blob_size = input_blob_size;
+
     for (auto pp : preprocessors) {
         if (!pp)
             break;
-        pp->preprocess(original_blob, storage_blob, query_blob, input_blob_size, this->alignment);
+        pp->preprocess(original_blob, storage_blob, query_blob, storage_blob_size, query_blob_size,
+                       this->alignment);
     }
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/spaces/computer/preprocessor_container.h]: we have storedDataSize in the index defintion

**Review by meiravgri — APPROVED** [review]

File: src/VecSim/spaces/computer/preprocessors.h [hunk 6]
```diff
@@ -119,3 +139,215 @@ class CosinePreprocessor : public PreprocessorInterface {
...
+ *              + delta_x * delta_y * Σ(qx_i * qy_i)
+ *   where:
+ *     - sum_x, sum_y are precomputed sums of original values
+ *     - Σqx_i = (sum_x - dim * min_x) / delta_x  (sum of quantized values, derived from stored sum)
+ *     - Σqy_i = (sum_y - dim * min_y) / delta_y
+ *
+ * For L2:
+ *   ||x - y||² = sum_sq_x + sum_sq_y - 2 * IP(x, y)
+ *   where sum_sq_x, sum_sq_y are precomputed sums of squared original values.
+ */
```
Comment (dor-forer) [inline src/VecSim/spaces/computer/preprocessors.h]: It's very bdidim, I like it

Comment (meiravgri) [reply src/VecSim/spaces/computer/preprocessors.h]: but does your ai like it?

**Review by meiravgri — APPROVED** [review]
